File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/XmlDeclarationSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class XmlDeclarationSyntax : VisualBasicSyntaxNode
	{
		internal readonly PunctuationSyntax _lessThanQuestionToken;

		internal readonly KeywordSyntax _xmlKeyword;

		internal readonly XmlDeclarationOptionSyntax _version;

		internal readonly XmlDeclarationOptionSyntax _encoding;

		internal readonly XmlDeclarationOptionSyntax _standalone;

		internal readonly PunctuationSyntax _questionGreaterThanToken;

		internal static Func<ObjectReader, object> CreateInstance;

		internal PunctuationSyntax LessThanQuestionToken => _lessThanQuestionToken;

		internal KeywordSyntax XmlKeyword => _xmlKeyword;

		internal XmlDeclarationOptionSyntax Version => _version;

		internal XmlDeclarationOptionSyntax Encoding => _encoding;

		internal XmlDeclarationOptionSyntax Standalone => _standalone;

		internal PunctuationSyntax QuestionGreaterThanToken => _questionGreaterThanToken;

		internal XmlDeclarationSyntax(SyntaxKind kind, PunctuationSyntax lessThanQuestionToken, KeywordSyntax xmlKeyword, XmlDeclarationOptionSyntax version, XmlDeclarationOptionSyntax encoding, XmlDeclarationOptionSyntax standalone, PunctuationSyntax questionGreaterThanToken)
			: base(kind)
		{
			base._slotCount = 6;
			AdjustFlagsAndWidth(lessThanQuestionToken);
			_lessThanQuestionToken = lessThanQuestionToken;
			AdjustFlagsAndWidth(xmlKeyword);
			_xmlKeyword = xmlKeyword;
			AdjustFlagsAndWidth(version);
			_version = version;
			if (encoding != null)
			{
				AdjustFlagsAndWidth(encoding);
				_encoding = encoding;
			}
			if (standalone != null)
			{
				AdjustFlagsAndWidth(standalone);
				_standalone = standalone;
			}
			AdjustFlagsAndWidth(questionGreaterThanToken);
			_questionGreaterThanToken = questionGreaterThanToken;
		}

		internal XmlDeclarationSyntax(SyntaxKind kind, PunctuationSyntax lessThanQuestionToken, KeywordSyntax xmlKeyword, XmlDeclarationOptionSyntax version, XmlDeclarationOptionSyntax encoding, XmlDeclarationOptionSyntax standalone, PunctuationSyntax questionGreaterThanToken, ISyntaxFactoryContext context)
			: base(kind)
		{
			base._slotCount = 6;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(lessThanQuestionToken);
			_lessThanQuestionToken = lessThanQuestionToken;
			AdjustFlagsAndWidth(xmlKeyword);
			_xmlKeyword = xmlKeyword;
			AdjustFlagsAndWidth(version);
			_version = version;
			if (encoding != null)
			{
				AdjustFlagsAndWidth(encoding);
				_encoding = encoding;
			}
			if (standalone != null)
			{
				AdjustFlagsAndWidth(standalone);
				_standalone = standalone;
			}
			AdjustFlagsAndWidth(questionGreaterThanToken);
			_questionGreaterThanToken = questionGreaterThanToken;
		}

		internal XmlDeclarationSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, PunctuationSyntax lessThanQuestionToken, KeywordSyntax xmlKeyword, XmlDeclarationOptionSyntax version, XmlDeclarationOptionSyntax encoding, XmlDeclarationOptionSyntax standalone, PunctuationSyntax questionGreaterThanToken)
			: base(kind, errors, annotations)
		{
			base._slotCount = 6;
			AdjustFlagsAndWidth(lessThanQuestionToken);
			_lessThanQuestionToken = lessThanQuestionToken;
			AdjustFlagsAndWidth(xmlKeyword);
			_xmlKeyword = xmlKeyword;
			AdjustFlagsAndWidth(version);
			_version = version;
			if (encoding != null)
			{
				AdjustFlagsAndWidth(encoding);
				_encoding = encoding;
			}
			if (standalone != null)
			{
				AdjustFlagsAndWidth(standalone);
				_standalone = standalone;
			}
			AdjustFlagsAndWidth(questionGreaterThanToken);
			_questionGreaterThanToken = questionGreaterThanToken;
		}

		internal XmlDeclarationSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 6;
			PunctuationSyntax punctuationSyntax = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax);
				_lessThanQuestionToken = punctuationSyntax;
			}
			KeywordSyntax keywordSyntax = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax != null)
			{
				AdjustFlagsAndWidth(keywordSyntax);
				_xmlKeyword = keywordSyntax;
			}
			XmlDeclarationOptionSyntax xmlDeclarationOptionSyntax = (XmlDeclarationOptionSyntax)reader.ReadValue();
			if (xmlDeclarationOptionSyntax != null)
			{
				AdjustFlagsAndWidth(xmlDeclarationOptionSyntax);
				_version = xmlDeclarationOptionSyntax;
			}
			XmlDeclarationOptionSyntax xmlDeclarationOptionSyntax2 = (XmlDeclarationOptionSyntax)reader.ReadValue();
			if (xmlDeclarationOptionSyntax2 != null)
			{
				AdjustFlagsAndWidth(xmlDeclarationOptionSyntax2);
				_encoding = xmlDeclarationOptionSyntax2;
			}
			XmlDeclarationOptionSyntax xmlDeclarationOptionSyntax3 = (XmlDeclarationOptionSyntax)reader.ReadValue();
			if (xmlDeclarationOptionSyntax3 != null)
			{
				AdjustFlagsAndWidth(xmlDeclarationOptionSyntax3);
				_standalone = xmlDeclarationOptionSyntax3;
			}
			PunctuationSyntax punctuationSyntax2 = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax2 != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax2);
				_questionGreaterThanToken = punctuationSyntax2;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_lessThanQuestionToken);
			writer.WriteValue(_xmlKeyword);
			writer.WriteValue(_version);
			writer.WriteValue(_encoding);
			writer.WriteValue(_standalone);
			writer.WriteValue(_questionGreaterThanToken);
		}

		static XmlDeclarationSyntax()
		{
			CreateInstance = (ObjectReader o) => new XmlDeclarationSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(XmlDeclarationSyntax), (ObjectReader r) => new XmlDeclarationSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _lessThanQuestionToken, 
				1 => _xmlKeyword, 
				2 => _version, 
				3 => _encoding, 
				4 => _standalone, 
				5 => _questionGreaterThanToken, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new XmlDeclarationSyntax(base.Kind, newErrors, GetAnnotations(), _lessThanQuestionToken, _xmlKeyword, _version, _encoding, _standalone, _questionGreaterThanToken);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new XmlDeclarationSyntax(base.Kind, GetDiagnostics(), annotations, _lessThanQuestionToken, _xmlKeyword, _version, _encoding, _standalone, _questionGreaterThanToken);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitXmlDeclaration(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/XmlDeclarationSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class XmlDeclarationSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
		Friend ReadOnly _lessThanQuestionToken As PunctuationSyntax

		Friend ReadOnly _xmlKeyword As KeywordSyntax

		Friend ReadOnly _version As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax

		Friend ReadOnly _encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax

		Friend ReadOnly _standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax

		Friend ReadOnly _questionGreaterThanToken As PunctuationSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property Encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax
			Get
				Return Me._encoding
			End Get
		End Property

		Friend ReadOnly Property LessThanQuestionToken As PunctuationSyntax
			Get
				Return Me._lessThanQuestionToken
			End Get
		End Property

		Friend ReadOnly Property QuestionGreaterThanToken As PunctuationSyntax
			Get
				Return Me._questionGreaterThanToken
			End Get
		End Property

		Friend ReadOnly Property Standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax
			Get
				Return Me._standalone
			End Get
		End Property

		Friend ReadOnly Property Version As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax
			Get
				Return Me._version
			End Get
		End Property

		Friend ReadOnly Property XmlKeyword As KeywordSyntax
			Get
				Return Me._xmlKeyword
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal lessThanQuestionToken As PunctuationSyntax, ByVal xmlKeyword As KeywordSyntax, ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal questionGreaterThanToken As PunctuationSyntax)
			MyBase.New(kind)
			MyBase._slotCount = 6
			MyBase.AdjustFlagsAndWidth(lessThanQuestionToken)
			Me._lessThanQuestionToken = lessThanQuestionToken
			MyBase.AdjustFlagsAndWidth(xmlKeyword)
			Me._xmlKeyword = xmlKeyword
			MyBase.AdjustFlagsAndWidth(version)
			Me._version = version
			If (encoding IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(encoding)
				Me._encoding = encoding
			End If
			If (standalone IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(standalone)
				Me._standalone = standalone
			End If
			MyBase.AdjustFlagsAndWidth(questionGreaterThanToken)
			Me._questionGreaterThanToken = questionGreaterThanToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal lessThanQuestionToken As PunctuationSyntax, ByVal xmlKeyword As KeywordSyntax, ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal questionGreaterThanToken As PunctuationSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind)
			MyBase._slotCount = 6
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(lessThanQuestionToken)
			Me._lessThanQuestionToken = lessThanQuestionToken
			MyBase.AdjustFlagsAndWidth(xmlKeyword)
			Me._xmlKeyword = xmlKeyword
			MyBase.AdjustFlagsAndWidth(version)
			Me._version = version
			If (encoding IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(encoding)
				Me._encoding = encoding
			End If
			If (standalone IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(standalone)
				Me._standalone = standalone
			End If
			MyBase.AdjustFlagsAndWidth(questionGreaterThanToken)
			Me._questionGreaterThanToken = questionGreaterThanToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal lessThanQuestionToken As PunctuationSyntax, ByVal xmlKeyword As KeywordSyntax, ByVal version As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal encoding As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal standalone As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax, ByVal questionGreaterThanToken As PunctuationSyntax)
			MyBase.New(kind, errors, annotations)
			MyBase._slotCount = 6
			MyBase.AdjustFlagsAndWidth(lessThanQuestionToken)
			Me._lessThanQuestionToken = lessThanQuestionToken
			MyBase.AdjustFlagsAndWidth(xmlKeyword)
			Me._xmlKeyword = xmlKeyword
			MyBase.AdjustFlagsAndWidth(version)
			Me._version = version
			If (encoding IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(encoding)
				Me._encoding = encoding
			End If
			If (standalone IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(standalone)
				Me._standalone = standalone
			End If
			MyBase.AdjustFlagsAndWidth(questionGreaterThanToken)
			Me._questionGreaterThanToken = questionGreaterThanToken
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 6
			Dim punctuationSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax)
				Me._lessThanQuestionToken = punctuationSyntax
			End If
			Dim keywordSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax)
				Me._xmlKeyword = keywordSyntax
			End If
			Dim xmlDeclarationOptionSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)
			If (xmlDeclarationOptionSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(xmlDeclarationOptionSyntax)
				Me._version = xmlDeclarationOptionSyntax
			End If
			Dim xmlDeclarationOptionSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)
			If (xmlDeclarationOptionSyntax1 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(xmlDeclarationOptionSyntax1)
				Me._encoding = xmlDeclarationOptionSyntax1
			End If
			Dim xmlDeclarationOptionSyntax2 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationOptionSyntax)
			If (xmlDeclarationOptionSyntax2 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(xmlDeclarationOptionSyntax2)
				Me._standalone = xmlDeclarationOptionSyntax2
			End If
			Dim punctuationSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax1 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax1)
				Me._questionGreaterThanToken = punctuationSyntax1
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitXmlDeclaration(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.XmlDeclarationSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Select Case i
				Case 0
					greenNode = Me._lessThanQuestionToken
					Exit Select
				Case 1
					greenNode = Me._xmlKeyword
					Exit Select
				Case 2
					greenNode = Me._version
					Exit Select
				Case 3
					greenNode = Me._encoding
					Exit Select
				Case 4
					greenNode = Me._standalone
					Exit Select
				Case 5
					greenNode = Me._questionGreaterThanToken
					Exit Select
				Case Else
					greenNode = Nothing
					Exit Select
			End Select
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._lessThanQuestionToken, Me._xmlKeyword, Me._version, Me._encoding, Me._standalone, Me._questionGreaterThanToken)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.XmlDeclarationSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._lessThanQuestionToken, Me._xmlKeyword, Me._version, Me._encoding, Me._standalone, Me._questionGreaterThanToken)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._lessThanQuestionToken, IObjectWritable))
			writer.WriteValue(DirectCast(Me._xmlKeyword, IObjectWritable))
			writer.WriteValue(DirectCast(Me._version, IObjectWritable))
			writer.WriteValue(DirectCast(Me._encoding, IObjectWritable))
			writer.WriteValue(DirectCast(Me._standalone, IObjectWritable))
			writer.WriteValue(DirectCast(Me._questionGreaterThanToken, IObjectWritable))
		End Sub
	End Class
End Namespace